### ISSUE TITLE
very high eval loss with gemma3 with sample packing on and eval packing off

### ISSUE THREAD

**Issue by sageof6path** [axolotl-ai-cloud/axolotl#3465]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Eval loss should be identical if eval sample packing is enabled vs disabled

### Current behaviour

Eval loss is very high when eval sample packing is disabled but when we enable eval sample packing it is coming low and the gap between them is very high.

### Steps to reproduce

You need to run the config provided below.

1) Run the config with sample packing enabled but eval packing is disabled. you will get high eval loss similar to the logs in https://github.com/sageof6path/axolotl/blob/main/examples/gemma3/high_eval_loss_logs.txt

2) Next, keep sample packing enabled but disable eval sample packing. you should get low eval loss. similar to logs in https://github.com/sageof6path/axolotl/blob/main/examples/gemma3/low_eval_loss.txt

### Config yaml

```yaml
base_model: google/gemma-3-1b-it

# gemma3 doesn't seem to play nice with ddp
ddp_find_unused_parameters: true
experimental_skip_move_to_device: true  # prevent OOM by NOT putting model to GPU before sharding

plugins:
- axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
- axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true

chat_template: gemma3
eot_tokens:
- <end_of_turn>

dataset_prepared_path: last_run_prepared
output_dir: output

datasets:
  - path: philschmid/dolly-15k-oai-style
    type: chat_template
    split: train[:90%]
    chat_template: gemma3
    field_messages: messages

test_datasets:
  - path: philschmid/dolly-15k-oai-style
    type: chat_template
    split: train[90%:]
    chat_template: gemma3
    field_messages: messages

flash_attention: true

sequence_len: 8192
sample_packing: true
eval_sample_packing: false

use_kernels: true

micro_batch_size: 2
gradient_accumulation_steps: 1
num_epochs: 3
optimizer: adamw_torch_fused
learning_rate: 2e-4

lr_scheduler: cosine

bf16: true
tf32: true
logging_steps: 1

gradient_checkpointing: true
activation_offloading: true

sample_packing_bin_size: 200
evals_per_epoch: 4
save_strategy: 'no'
include_tokens_per_second: true
save_safetensors: true
use_tensorboard: true
```

### Possible solution

Not sure but may be problem lies here: https://github.com/axolotl-ai-cloud/axolotl/blob/2b6f4a6c9b9acccf9eb1c9f4e730a3a2f25b9488/src/axolotl/utils/trainer.py#L253

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

using latest axolotl docker

### axolotl branch-commit

axolotlai/axolotl:main-latest

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by ved1beta** [comment]
hey thanks for reporting this please can you try this brach https://github.com/axolotl-ai-cloud/axolotl/pull/3478

**Comment by iambbking** [comment]
I have seen something similar. In order to unblock myself, I set `flash_attention` to `false`, and then the loss on both the training and evaluation look reasonable.

**Comment by NanoCode012** [comment]
> I set `flash_attention` to `false`, and then the loss on both the training and evaluation look reasonable.

That could be an indicator relating to packing . Could you give the linked PR a try?

**Comment by NanoCode012** [comment]
We merged the related branch, so will close this PR. Let us know if that didn't fix it
